bittensor/_dataset/dataset_impl.py

```diff
@@ -1,538 +1,494 @@
+        self.max_threads = max_threads
+        self._idle_semaphore = threading.Semaphore(0)
+        self._threads = []
+        self._shutdown_lock = threading.Lock()
+        self._shutdown = False
+
+
+    def submit(self, fn, args:Optional[list]=[],kwargs:Optional[dict]={}) -> Any:
+        '''
+        Submit a function with args and kwargs on a seperate thread.
+
+        Args
+            fn (Callable):
+                Function to place on the thread.
+            args (list):
+                Arguments to a function.
+            kwargs (dict):
+                Key word arguments to a function.
+        '''
+        with self._shutdown_lock:
+            if self._shutdown:
+                raise RuntimeError('cannot schedule new futures after shutdown')
+            
+            thread = threading.Thread(target=fn, args=args, kwargs=kwargs, daemon=True)
+            thread.start()
+            self._threads.append(thread)
 
-        except Exception as E:
-            logger.error(f"Failed to get from IPFS {file_meta['Name']} {E}")
-            return None
+        return thread
 
-        return response
 
-    def __len__(self):
-        """ Returns length of the dataset that the dataset is processing
-        """
+    @property
+    def threads(self):
+        '''List threads.'''
+        return self._threads
 
-    def __getitem__(self, idx):
-        """ Returns the next batch from the dataset.
-        """
+    def __del__(self):
+        self.shutdown()
 
-class GenesisTextDataset( Dataset ):
-    """ One kind of dataset that caters for the data from ipfs 
+    def shutdown(self, wait=True):
+        '''Shutdown threads'''
+        if wait:
+            for t in self._threads:
+                t.join()
+
+
+
+
+class GenesisTextDataset:
+    """ Implementation for the dataset class, which handles dataloading from ipfs
     """
+
+    ipfs_url = 'http://global.ipfs.opentensor.ai/api/v0'
+    dataset_dir = 'http://global.ipfs.opentensor.ai/api/v0/cat' 
+    text_dir = 'http://global.ipfs.opentensor.ai/api/v0/object/get'
+    mountain_hash = 'QmSdDg6V9dgpdAFtActs75Qfc36qJtm9y8a7yrQ1rHm7ZX'
+
     def __init__(
-        self,
-        block_size,
-        batch_size,
-        num_workers,
-        dataset_name,
-        data_dir,
-        save_dataset,
-        max_datasets,
-        no_tokenizer, 
-        num_batches,
-        max_directories
-    ):
-        super().__init__()
-        self.block_size = block_size
+            self, 
+            batch_size: int=8, 
+            sequence_length: int=256,
+            block_size: int = 10000,
+            num_workers: int = 1,
+            datasets: Optional[Union[List[str], str]]=['ArXiv'], 
+            loop:'asyncio.loop'=None, 
+            no_tokenizer: bool = False,
+            data_dir: str =  os.path.expanduser('~/./bittensor/data'),
+            save_dataset : bool = False,
+            load_dataset : bool = True,
+            run_generator:bool=False,
```

Comment:
You should write a unit test where `run_generator is True`. Both as proof it works as intended and to ensure nobody else inadvertently breaks it in the future.